Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -203,146 +203,180 @@ function xASL_wrp_ResampleASL(x)
-		PWIsingle = xASL_stat_MeanNan(ASL_im,4); % Average across PLDs
+    
+    % =====================================================================
+    % A) Check subtraction
+    if numel(size(ASL_im))>4
+        warning('In BIDS ASL NIfTIs should have [X Y Z n/PLD/TE/etc] as 4 dimensions');
+        error(['NIfTI has more than 4 dimensions: ' PathASL4D{iSpace}]);
         
-		% Save single PWI
-		fprintf('%s\n', PathPWI{iSpace});
-		xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWIsingle, 32, false);
-					
-    elseif round(dim4/2)~=dim4/2
-        warning('Odd number of control-label pairs, skipping');
+    elseif dim4>1 && round(dim4/2)~=dim4/2
+        warning('Odd (i.e., not even) number of control-label pairs, skipping');
         return;
         
+        
+    % =====================================================================
+    % B) Time-Encoded subtraction
     elseif x.modules.asl.bTimeEncoded
+        
+        %% 1) Create PWI4D
         % Decoding of TimeEncoded data - it outputs a decoded image and it also saves as a NII
-        ASL_im = xASL_quant_HadamardDecoding(PathASL{iSpace}, x.Q);
+        PWI4D = xASL_quant_HadamardDecoding(PathASL4D{iSpace}, x.Q);
 		
-		% Calculate Hadamard block size (number of unique PLDs * multiTEs) = number of volumes per block
-        vectorSizePLD = length(unique(x.Q.Initial_PLD));
-        blockSize = x.Q.NumberEchoTimes * (vectorSizePLD - (vectorSizePLD/x.Q.TimeEncodedMatrixSize)); 
-		
-        if blockSize ~= size(ASL_im,4) % In case there is more than 1 block, we will average across repetitions
-			if mod(size(ASL_im,4), blockSize)
-				% Number of total volumes cannot be divided by blockSize
-				error(['Total number of volumes ' xASL_num2str(size(ASL_im,4)) ' cannot be composed of blocks of size ' xASL_num2str(blockSize)]);
-			else
-				PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), blockSize); % preallocate PWI
-				for iBlock = 1:blockSize
-					PWI(:,:,:,iBlock) = xASL_stat_MeanNan(ASL_im(:,:,:,iBlock:blockSize:end), 4); % Averaged PWI4D across repetitions
-				end
-			end
-        else % if Hadamard Block size == unique PLDs * number of multi-TEs -> a single repetitions -> we don't want to average
-            PWI = ASL_im;
+        %% 2) Create PWI3D
+		% Calculate Hadamard block size (number of unique PLDs.*labeling durations.* echo times) = number of volumes per repetition
+        unique_InitialPLD = unique(x.Q.Initial_PLD);
+        nUnique_InitialPLD = length(unique_InitialPLD);
+
+        nEchoTimes = length(uniquetol(x.EchoTime,0.001)); % Obtain the number of echo times
+        % We do this here now for each sequence, but we could also do this
+        % at the start of the ASL module
+
+        nVolumesPerRepetition = nEchoTimes .* (nUnique_InitialPLD - (nUnique_InitialPLD./x.Q.TimeEncodedMatrixSize));
+        % nVolumesPerRepetition is also called blockSize by some
+		nRepetitions = size(PWI4D,4) ./ nVolumesPerRepetition;

```

Comment:
Please do not remove simple code (mod) to replace it by complicated code (division + round). Modulo is extremely common and simply function made for checking the division so that you don't have to use this rounding.

Also - please revert the comments. You cannot say that you have N extra volumes or that you are missing N volumes. This cannot be said for certainty and is just a wrong comment. You can only say that the volumes match. Please revert the entire edit.